### PR TITLE
Add proto files along with linter checks

### PR DESCRIPTION
**Pull request by prasanna08**
This PR adds proto files necessary for communication between Oppia and Oppia-ml along with lint tests and GitHub actions workflow for linter tests.

### CODE REVIEW

**Comment by prasanna08** [comment]
Hi @BenHenning @seanlip PTAL.

File: install_prototool.sh [hunk 1]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env bash
```
Comment (seanlip) [inline install_prototool.sh]: Can you use python scripts instead? We are basically getting rid of .sh from the entire codebase.

**Review by seanlip — CHANGES_REQUESTED** [review]

File: install_prototool.sh (hunk 1 above)
Comment (prasanna08) [reply install_prototool.sh]: Sure.

Comment (prasanna08) [reply install_prototool.sh]: Done

File: install_prototool.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    'prototool-Linux-x86_64')
+
+
+def download_file(source_url, target_dir, filename):
+    """Downloads a file and save it to a given directory.
+
+    File is downloaded only if it does not already exist.
+
+    Args:
+      source_url: the URL of the file.
```
Comment (seanlip) [inline install_prototool.py]: Let's follow existing code conventions (see oppia codebase)

- indent by 4
- include typeinfo
- start descriptions with capital letters

File: pre_commit_linter.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+list of lint errors to the terminal. If the directory path is passed,
+it will lint all Python files in that directory; otherwise,
+it will only lint files that have been touched in this commit.
+
+IMPORTANT NOTES:
+
+1.  Before running this script, you must install third-party dependencies by
+    running
+
+        bash install_prototool.sh
```
Comment (seanlip) [inline pre_commit_linter.py]: Please fix.

File: pre_commit_linter.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+_MESSAGE_TYPE_SUCCESS = 'SUCCESS'
+_MESSAGE_TYPE_FAILED = 'FAILED'
+
+EXCLUDED_PATHS = ('third_party/*',)
+
+PROTOTOOL_CONFIG_FILE = 'prototool_config.json'
+
+def _get_changed_filenames():
+    """Returns a list of modified files (both staged and unstaged)
```
Comment (seanlip) [inline pre_commit_linter.py]: End docstrings with periods.

File: pre_commit_linter.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+
+    Returns:
+        a list of filenames of modified files.
```
Comment (seanlip) [inline pre_commit_linter.py]: Missing typeinfo. Start descriptions with capital letters. Ditto below.

Btw -- this is unrelated, but perhaps worth adding a python linter?

File: pre_commit_linter.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+
+def _lint_proto_files(files_to_lint, result, config):
+    """Prints a list of lint errors in the given list of Proto files.
+
+    Args:
+        files_to_lint: list(str). A list of filepaths to lint.
+        result: multiprocessing.Queue. A queue to put results of test.
+    """
+    start_time = time.time()
+    are_there_errors = False
```
Comment (seanlip) [inline pre_commit_linter.py]: errors_exist (in Oppia codebases, variables never start with verbs, and methods/functions always do)

File: pre_commit_linter.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+
+
+def _lint_proto_files(files_to_lint, result, config):
+    """Prints a list of lint errors in the given list of Proto files.
+
+    Args:
+        files_to_lint: list(str). A list of filepaths to lint.
+        result: multiprocessing.Queue. A queue to put results of test.
+    """
+    start_time = time.time()
```
Comment (seanlip) [inline pre_commit_linter.py]: Perhaps, conceptually, this should be moved lower down, after you've checked whether there are files to lint.

File: pre_commit_linter.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+    if are_there_errors:
+        result.put('%s    Proto linting failed' % _MESSAGE_TYPE_FAILED)
+    else:
+        result.put('%s   %s Proto files linted (%.1f secs)' % (
+            _MESSAGE_TYPE_SUCCESS, num_proto_files, time.time() - start_time))
+
+    print('Proto linting finished.')
+
+def _get_all_files():
+    """This function is used to check if this script is ran from
```
Comment (seanlip) [inline pre_commit_linter.py]: ran --> run

File: pre_commit_linter.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    elif parsed_args.files:
+        valid_filepaths = []
+        invalid_filepaths = []
+        for f in parsed_args.files:
+            if os.path.isfile(f):
+                valid_filepaths.append(f)
+            else:
+                invalid_filepaths.append(f)
+        if invalid_filepaths:
+            print('The following file(s) do not exist: %s\n'
```
Comment (seanlip) [inline pre_commit_linter.py]: Prefer breaking after '(' -- see https://github.com/oppia/oppia/wiki/Coding-style-guide#python

**Review by seanlip — CHANGES_REQUESTED** [review]
Thanks! Took a full pass but basically looks good.

**Comment by seanlip** [comment]
Also I don't see the github actions workflow being triggered on this PR. It should show up in the list of status checks. Could you check this?

File: pre_commit_linter.py (hunk 5 above)
Comment (prasanna08) [reply pre_commit_linter.py]: I don't think that a python linter is necessary as the repo would contain only `.proto` files, except for linter and installer scripts. There won't be any third party usable python code in this repo.

File: install_prototool.py (hunk 2 above)
Comment (prasanna08) [reply install_prototool.py]: Done

File: pre_commit_linter.py (hunk 3 above)
Comment (prasanna08) [reply pre_commit_linter.py]: Done

File: pre_commit_linter.py (hunk 4 above)
Comment (prasanna08) [reply pre_commit_linter.py]: Done

File: pre_commit_linter.py (hunk 5 above)
Comment (prasanna08) [reply pre_commit_linter.py]: Done

File: pre_commit_linter.py (hunk 6 above)
Comment (prasanna08) [reply pre_commit_linter.py]: Done

File: pre_commit_linter.py (hunk 7 above)
Comment (prasanna08) [reply pre_commit_linter.py]: Done

File: pre_commit_linter.py (hunk 8 above)
Comment (prasanna08) [reply pre_commit_linter.py]: Done

File: pre_commit_linter.py (hunk 9 above)
Comment (prasanna08) [reply pre_commit_linter.py]: Done

**Comment by prasanna08** [comment]
I think workflow is not executed because it is not the develop branch, yet.

**Comment by prasanna08** [comment]
Hi @seanlip, I was able to run the github actions workflow on my local develop branch, post merging this PR's branch with it. The tests are running [successfully](https://github.com/prasanna08/oppia-ml-proto/actions). I had to fix a couple of strings in code but now it is running fine.

Hence, I think after merging this PR, tests will run on develop branch and all future incoming PRs.

**EDIT**: I checked again, actions are only triggered when the workflow code is merged. I will fix my local develop branch by force pushing from oppia-ml-proto when this PR is merged.

**Comment by BenHenning** [comment]
To keep things clean @prasanna08 maybe split this up between: adding the protos, precomit hook, gitignore, etc. and adding support for the GitHub actions linter. I actually think you can add the actions checker first, then update this PR after that's merged in & the check is enabled to verify it's passing.

**Review by BenHenning — APPROVED** [review]
Thanks @prasanna08! I looked at this after Sean, but it LGTM other than my comment above (please add the GitHub actions check first, then this to verify it's passing the lint checks).

File: pre_commit_linter.py [hunk 10]
```diff
@@ -83,15 +83,15 @@
...
 
 PROTOTOOL_CONFIG_FILE = 'prototool_config.json'
 
 def _get_changed_filenames():
-    """Returns a list of modified files (both staged and unstaged)
+    """Returns a list of modified files (both staged and unstaged).
 
     Returns:
-        a list of filenames of modified files.
+        List(str). A list of filenames of modified files.
```
Comment (seanlip) [inline pre_commit_linter.py:94]: list(str).

Ditto elsewhere -- types should not be capitalized. Just the description after that.

File: pre_commit_linter.py [hunk 11]
```diff
@@ -110,7 +110,8 @@ def _get_all_files_in_directory(dir_path):
         dir_path: str. Path to the folder to be linted.
 
     Returns:
-        a list of files in directory and subdirectories without excluded files.
+        List(str). a list of files in directory and subdirectories without
```
Comment (seanlip) [inline pre_commit_linter.py:113]: a --> A

**Review by seanlip — CHANGES_REQUESTED** [review]
Only small comments, otherwise LGTM. Deferring final approval until after you've done what @BenHenning suggests (i.e. the GitHub actions check goes in first) -- thanks!

File: pre_commit_linter.py (hunk 10 above)
Comment (prasanna08) [reply pre_commit_linter.py:94]: Done in #2.

File: pre_commit_linter.py (hunk 11 above)
Comment (prasanna08) [reply pre_commit_linter.py:113]: Done in #2

**Comment by prasanna08** [comment]
Closing this as the work will be split in two separate PRs. #2 adds workflow for linter checks. Once it is merged, I will add proto files.

**Pull request closed without merging**